File: services/attendance-service/src/test/java/com/visor/school/attendance/model/AttendanceSessionTest.java
package com.visor.school.attendance.model;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.time.LocalDate;
import java.util.UUID;

class AttendanceSessionTest {

    @Test
    void shouldCreateAttendanceSessionWithRequiredFields() {
        UUID classId = UUID.randomUUID();
        UUID createdBy = UUID.randomUUID();
        LocalDate date = LocalDate.now();

        AttendanceSession session = new AttendanceSession(classId, date, createdBy);
        // DelegatedTo is not in the constructor, set via method or left null
        
        assertNull(session.getId());
        assertEquals(AttendanceSessionStatus.PENDING, session.getStatus());
        assertNull(session.getDelegatedTo());
        assertNotNull(session.getCreatedBy());
        assertNull(session.getApprovedBy());
        assertNull(session.getRejectedBy());
    }

    @Test
    void shouldThrowExceptionForFutureDate() {
        assertThrows(IllegalArgumentException.class, () -> {
            new AttendanceSession(
                UUID.randomUUID(),
                LocalDate.now().plusDays(1),
                UUID.randomUUID()
            );
        });
    }

    @Test
    void shouldMarkSessionAsCollected() {
        AttendanceSession session = new AttendanceSession(
            UUID.randomUUID(),
            LocalDate.now(),
            UUID.randomUUID()
        );
        // Delegate first as required by markAsCollected
        session.delegateTo(UUID.randomUUID());

        session.markAsCollected();

        assertEquals(AttendanceSessionStatus.COLLECTED, session.getStatus());
        assertNotNull(session.getCollectedAt());
    }

    @Test
    void shouldThrowExceptionWhenMarkingCollectedWithoutDelegation() {
        AttendanceSession session = new AttendanceSession(
            UUID.randomUUID(),
            LocalDate.now(),
            UUID.randomUUID()
        );
        // Not delegated

        assertThrows(IllegalStateException.class, session::markAsCollected);
    }

    @Test
    void shouldApproveSession() {
        UUID teacherId = UUID.randomUUID();
        AttendanceSession session = new AttendanceSession(
            UUID.randomUUID(),
            LocalDate.now(),
            UUID.randomUUID()
        );
        session.delegateTo(UUID.randomUUID());
        session.markAsCollected();

        session.approve(teacherId);

        assertEquals(AttendanceSessionStatus.APPROVED, session.getStatus());
        assertEquals(teacherId, session.getApprovedBy());
        assertNotNull(session.getApprovedAt());
        assertNull(session.getRejectedBy());
        assertNull(session.getRejectionReason());
    }

    @Test
    void shouldThrowExceptionWhenApprovingNonCollectedSession() {
        AttendanceSession session = new AttendanceSession(
            UUID.randomUUID(),
            LocalDate.now(),
            UUID.randomUUID()
        );
        
        assertThrows(IllegalStateException.class, () -> session.approve(UUID.randomUUID()));
    }

    @Test
    void shouldRejectSessionWithReason() {
        UUID teacherId = UUID.randomUUID();
        AttendanceSession session = new AttendanceSession(
            UUID.randomUUID(),
            LocalDate.now(),
            UUID.randomUUID()
        );
        session.delegateTo(UUID.randomUUID());
        session.markAsCollected();

        session.reject(teacherId, "Incomplete attendance data");

        assertEquals(AttendanceSessionStatus.REJECTED, session.getStatus());
        assertEquals(teacherId, session.getRejectedBy());
        assertEquals("Incomplete attendance data", session.getRejectionReason());
        assertNotNull(session.getRejectedAt());
        assertNull(session.getCollectedAt()); // Reset after rejection
    }

    @Test
    void shouldThrowExceptionWhenRejectingWithoutReason() {
        AttendanceSession session = new AttendanceSession(
            UUID.randomUUID(),
            LocalDate.now(),
            UUID.randomUUID()
        );
        session.delegateTo(UUID.randomUUID());
        session.markAsCollected();

        assertThrows(IllegalArgumentException.class, () -> session.reject(UUID.randomUUID(), ""));
    }

    @Test
    void shouldResubmitRejectedSession() {
        AttendanceSession session = new AttendanceSession(
            UUID.randomUUID(),
            LocalDate.now(),
            UUID.randomUUID()
        );
        session.delegateTo(UUID.randomUUID());
        session.markAsCollected();
        session.reject(UUID.randomUUID(), "Incomplete");

        session.resubmit();

        assertEquals(AttendanceSessionStatus.COLLECTED, session.getStatus());
        assertNotNull(session.getCollectedAt());
        assertNull(session.getRejectionReason());
        assertNull(session.getRejectedBy());
    }

    @Test
    void shouldThrowExceptionWhenResubmittingNonRejectedSession() {
        AttendanceSession session = new AttendanceSession(
            UUID.randomUUID(),
            LocalDate.now(),
            UUID.randomUUID()
        );
        
        assertThrows(IllegalStateException.class, session::resubmit);
    }
}

File: services/attendance-service/src/test/kotlin/com/visor/school/attendanceservice/model/AttendanceSessionTest.kt
package com.visor.school.attendanceservice.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.time.LocalDate
import java.util.UUID

class AttendanceSessionTest {

    @Test
    fun `should create attendance session with required fields`() {
        val session = AttendanceSession(
            classId = UUID.randomUUID(),
            date = LocalDate.now(),
            delegatedTo = UUID.randomUUID(),
            createdBy = UUID.randomUUID()
        )

        assertNotNull(session.id)
        assertEquals(AttendanceSessionStatus.PENDING, session.status)
        assertNotNull(session.delegatedTo)
        assertNotNull(session.createdBy)
        assertNull(session.approvedBy)
        assertNull(session.rejectedBy)
    }

    @Test
    fun `should throw exception for future date`() {
        assertThrows<IllegalArgumentException> {
            AttendanceSession(
                classId = UUID.randomUUID(),
                date = LocalDate.now().plusDays(1),
                delegatedTo = UUID.randomUUID(),
                createdBy = UUID.randomUUID()
            )
        }
    }

    @Test
    fun `should mark session as collected`() {
        val session = AttendanceSession(
            classId = UUID.randomUUID(),
            date = LocalDate.now(),
            delegatedTo = UUID.randomUUID(),
            createdBy = UUID.randomUUID(),
            status = AttendanceSessionStatus.PENDING
        )

        session.markAsCollected()

        assertEquals(AttendanceSessionStatus.COLLECTED, session.status)
        assertNotNull(session.collectedAt)
    }

    @Test
    fun `should throw exception when marking collected without delegation`() {
        val session = AttendanceSession(
            classId = UUID.randomUUID(),
            date = LocalDate.now(),
            delegatedTo = UUID.randomUUID(),
            createdBy = UUID.randomUUID(),
            status = AttendanceSessionStatus.APPROVED
        )

        assertThrows<IllegalArgumentException> {
            session.markAsCollected()
        }
    }

    @Test
    fun `should approve session`() {
        val teacherId = UUID.randomUUID()
        val session = AttendanceSession(
            classId = UUID.randomUUID(),
            date = LocalDate.now(),
            delegatedTo = UUID.randomUUID(),
            createdBy = UUID.randomUUID(),
            status = AttendanceSessionStatus.COLLECTED
        )
        session.markAsCollected()

        session.approve(teacherId)

        assertEquals(AttendanceSessionStatus.APPROVED, session.status)
        assertEquals(teacherId, session.approvedBy)
        assertNotNull(session.approvedAt)
        assertNull(session.rejectedBy)
        assertNull(session.rejectionReason)
    }

    @Test
    fun `should throw exception when approving non-collected session`() {
        val session = AttendanceSession(
            classId = UUID.randomUUID(),
            date = LocalDate.now(),
            delegatedTo = UUID.randomUUID(),
            createdBy = UUID.randomUUID(),
            status = AttendanceSessionStatus.PENDING
        )

        assertThrows<IllegalArgumentException> {
            session.approve(UUID.randomUUID())
        }
    }

    @Test
    fun `should reject session with reason`() {
        val teacherId = UUID.randomUUID()
        val session = AttendanceSession(
            classId = UUID.randomUUID(),
            date = LocalDate.now(),
            delegatedTo = UUID.randomUUID(),
            createdBy = UUID.randomUUID(),
            status = AttendanceSessionStatus.COLLECTED
        )
        session.markAsCollected()

        session.reject(teacherId, "Incomplete attendance data")

        assertEquals(AttendanceSessionStatus.REJECTED, session.status)
        assertEquals(teacherId, session.rejectedBy)
        assertEquals("Incomplete attendance data", session.rejectionReason)
        assertNotNull(session.rejectedAt)
        assertNull(session.collectedAt) // Reset after rejection
    }

    @Test
    fun `should throw exception when rejecting without reason`() {
        val session = AttendanceSession(
            classId = UUID.randomUUID(),
            date = LocalDate.now(),
            delegatedTo = UUID.randomUUID(),
            createdBy = UUID.randomUUID(),
            status = AttendanceSessionStatus.COLLECTED
        )
        session.markAsCollected()

        assertThrows<IllegalArgumentException> {
            session.reject(UUID.randomUUID(), "")
        }
    }

    @Test
    fun `should resubmit rejected session`() {
        val session = AttendanceSession(
            classId = UUID.randomUUID(),
            date = LocalDate.now(),
            delegatedTo = UUID.randomUUID(),
            createdBy = UUID.randomUUID(),
            status = AttendanceSessionStatus.REJECTED
        )
        session.reject(UUID.randomUUID(), "Incomplete")

        session.resubmit()

        assertEquals(AttendanceSessionStatus.COLLECTED, session.status)
        assertNotNull(session.collectedAt)
        assertNull(session.rejectionReason)
        assertNull(session.rejectedBy)
    }

    @Test
    fun `should throw exception when resubmitting non-rejected session`() {
        val session = AttendanceSession(
            classId = UUID.randomUUID(),
            date = LocalDate.now(),
            delegatedTo = UUID.randomUUID(),
            createdBy = UUID.randomUUID(),
            status = AttendanceSessionStatus.PENDING
        )

        assertThrows<IllegalArgumentException> {
            session.resubmit()
        }
    }
}

